File: android/gemcore/src/main/kotlin/com/gemwallet/android/model/AssetData.kt
package com.gemwallet.android.model

import com.wallet.core.primitives.Account
import com.wallet.core.primitives.Asset
import com.wallet.core.primitives.AssetAssociation
import com.wallet.core.primitives.AssetMetaData
import com.wallet.core.primitives.Wallet
import com.wallet.core.primitives.WalletId

data class AssetData(
    val asset: Asset,
    val balance: AssetBalance = AssetBalance(asset),
    val account: Account,
    val walletId: WalletId,
    val price: AssetPriceInfo? = null,
    val metadata: AssetMetaData,
    val associations: List<AssetAssociation> = emptyList(),
) {
    fun toAssetInfo(): AssetInfo = AssetInfo(
        owner = account,
        asset = asset,
        balance = balance,
        walletId = walletId,
        price = price,
        metadata = metadata,
        associations = associations,
    )

    companion object {
        fun from(assetInfo: AssetInfo, wallet: Wallet, account: Account) = AssetData(
            asset = assetInfo.asset,
            account = account,
            walletId = wallet.id,
            balance = assetInfo.balance,
            price = assetInfo.price,
            metadata = assetInfo.metadata,
            associations = assetInfo.associations,
        )
    }
}

File: core/crates/primitives/src/asset_data.rs
use chrono::{DateTime, Utc};
use model_derive::Model;
use serde::{Deserialize, Serialize};

use crate::{Account, Asset, AssetAssociation, AssetMetaData, Balance, Price, PriceAlert};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Model)]
#[model(swift = "Equatable, Hashable, Sendable")]
#[serde(rename_all = "camelCase")]
pub struct AssetData {
    pub asset: Asset,
    pub balance: Balance,
    pub account: Account,
    pub price: Option<Price>,
    pub price_alerts: Vec<PriceAlert>,
    pub metadata: AssetMetaData,
    pub associations: Vec<AssetAssociation>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Model)]
#[model(swift = "Equatable, Hashable, Sendable")]
#[serde(rename_all = "camelCase")]
pub struct ChainAssetData {
    pub asset_data: AssetData,
    pub fee_asset_data: AssetData,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Model)]
#[model(swift = "Equatable, Hashable, Sendable")]
#[serde(rename_all = "camelCase")]
pub struct RecentAsset {
    pub asset: Asset,
    pub created_at: DateTime<Utc>,
}
